src/poprox_recommender/components/diversifiers/locality_calibration.py

```diff
@@ -25,13 +47,234 @@ def __call__(self, candidate_articles: ArticleSet, interest_profile: InterestPro
+    )
+
+    input_prompt = "News List: \n" + f"{news_list}" + "User Interest: \n" + user_interests
+    return gpt_generate(system_prompt, input_prompt)
+
+
+model = SentenceTransformer("all-MiniLM-L6-v2")
+
+
+def related_article(current_news_slug, current_topics, clicked_articles, k):
+    # TODO: do we need to consider the cold start?
+    target_embedding = model.encode([current_news_slug])[0]
+
+    subject_candidate_headlines = []
+    for idx in range(len(clicked_articles)):
+        slug = clicked_articles[int(idx)].subhead
+        # category = set([mention.entity.name for mention in clicked_articles[int(idx)].mentions])
+
+        # if any(item in current_topics for item in category):
+        subject_candidate_headlines.append(slug)
+
+    candidate_embeddings = model.encode(subject_candidate_headlines)
+    similarities = cosine_similarity([target_embedding], candidate_embeddings)[0]
+    top_k_indices = np.argsort(similarities)[-k:][::-1]
+
+    return top_k_indices
+
+
+def text_generation(candidate_articles, article_indices, interest_profile):
+    clicked_articles = interest_profile.clicked_articles.articles
+
+    recommended = []
+    for idx in article_indices:
+        news_list = []  # for context generation
+        current_news_slug = candidate_articles[int(idx)].subhead
+        current_topics = set([mention.entity.name for mention in candidate_articles[int(idx)].mentions])
+
+        temp_article = {}
+        temp_article["ID"] = "Main news"
+        temp_article["headline"] = candidate_articles[int(idx)].headline
+        temp_article["subhead"] = candidate_articles[int(idx)].subhead
+        temp_article["categories"] = set([mention.entity.name for mention in candidate_articles[int(idx)].mentions])
+        temp_article["topics"] = news_factor(
+            candidate_articles[int(idx)].headline, candidate_articles[int(idx)].subhead
+        )
+        news_list.append(temp_article)
+
+        # TODO: determine the k (3) value
+        clicked_articles_idx = related_article(current_news_slug, current_topics, clicked_articles, 3)
+
+        for idx_clicked in clicked_articles_idx:
+            temp_article = {}
+            temp_article["ID"] = "related articles"
+            temp_article["headline"] = clicked_articles[int(idx_clicked)].headline
+            temp_article["subhead"] = clicked_articles[int(idx_clicked)].subhead
+            temp_article["categories"] = set(
+                [mention.entity.name for mention in clicked_articles[int(idx_clicked)].mentions]
+            )
+            temp_article["topics"] = news_factor(
+                clicked_articles[int(idx_clicked)].headline, clicked_articles[int(idx_clicked)].subhead
+            )
+            news_list.append(temp_article)
+
+        candidate_articles[int(idx)].headline = generate_narrative(news_list, interest_profile.user_interests)
+        recommended.append(candidate_articles)
+    return recommended[0]
```

Comment:
I could be missing something, but it seems like this should return more than one article?
```suggestion
    return recommended
```